### PR TITLE
WS to manage Pull Requests

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+  private static void addPullRequest(ProjectPullRequests.ListWsResponse.Builder response, BranchDto branch, Map<String, BranchDto> mergeBranchesByUuid,
+    BranchStatistics branchStatistics, @Nullable String analysisDate) {
+    Optional<BranchDto> mergeBranch = Optional.ofNullable(mergeBranchesByUuid.get(branch.getMergeBranchUuid()));
+    ProjectPullRequests.PullRequest.Builder builder = ProjectPullRequests.PullRequest.newBuilder();
+    String branchKey = branch.getKey();
+    builder.setId(branchKey);
+    //TODO get the branch from PULL_REQUEST_DATA
+    builder.setBranch(branchKey);
+    //TODO get the URL from PULL_REQUEST_DATA
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fws-to-handle-pr&issues=AWGWV8TKjS02mAJE6Aan&open=AWGWV8TKjS02mAJE6Aan)

[ISSUE_KEY]: AWGWV8TKjS02mAJE6Aan

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    builder.setUrl("https://github.com/SonarSource/sonar-core-plugins/pull/32");
+    //TODO get the branch title from PULL_REQUEST_DATA
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fws-to-handle-pr&issues=AWGWV8TKjS02mAJE6Aao&open=AWGWV8TKjS02mAJE6Aao)

[ISSUE_KEY]: AWGWV8TKjS02mAJE6Aao

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    }
+  }
+
+  private static void addPullRequest(ProjectPullRequests.ListWsResponse.Builder response, BranchDto branch, Map<String, BranchDto> mergeBranchesByUuid,
+    BranchStatistics branchStatistics, @Nullable String analysisDate) {
+    Optional<BranchDto> mergeBranch = Optional.ofNullable(mergeBranchesByUuid.get(branch.getMergeBranchUuid()));
+    ProjectPullRequests.PullRequest.Builder builder = ProjectPullRequests.PullRequest.newBuilder();
+    String branchKey = branch.getKey();
+    builder.setId(branchKey);
+    //TODO get the branch from PULL_REQUEST_DATA
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fws-to-handle-pr&issues=AWGWV8TKjS02mAJE6Aam&open=AWGWV8TKjS02mAJE6Aam)

[ISSUE_KEY]: AWGWV8TKjS02mAJE6Aam

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java (hunk 2 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fws-to-handle-pr&issues=AWGWV8TKjS02mAJE6Aao&open=AWGWV8TKjS02mAJE6Aao)

[ISSUE_KEY]: AWGWV8TKjS02mAJE6Aao

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java (hunk 1 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fws-to-handle-pr&issues=AWGWV8TKjS02mAJE6Aan&open=AWGWV8TKjS02mAJE6Aan)

[ISSUE_KEY]: AWGWV8TKjS02mAJE6Aan

File: server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java (hunk 3 above)
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/branch/pr/ws/ListAction.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fws-to-handle-pr&issues=AWGWV8TKjS02mAJE6Aam&open=AWGWV8TKjS02mAJE6Aam)

[ISSUE_KEY]: AWGWV8TKjS02mAJE6Aam

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskFormatter.java [hunk 4]
```diff
@@ -144,12 +142,25 @@ public TaskFormatter(DbClient dbClient, System2 system2) {
...
+          .orElseThrow(() -> new IllegalStateException(format("Could not find branch type of task '%s'", taskUuid)));
+        builder.setBranchType(branchType);
+        switch (branchType) {
+          case LONG:
+          case SHORT:
+            builder.setBranch(b);
+            break;
+          case PULL_REQUEST:
+            builder.setPullRequest(b);
+            // TODO set pull request title
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskFormatter.java]: Code Smell Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fteryk%2Fws-to-handle-pr&issues=AWGWnltdRRutcW4uJrtr&open=AWGWnltdRRutcW4uJrtr)

[ISSUE_KEY]: AWGWnltdRRutcW4uJrtr

File: server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java [hunk 5]
```diff
@@ -97,30 +97,34 @@ public void handle(Request request, Response response) throws Exception {
       userSession.checkComponentPermission(UserRole.USER, project);
       checkArgument(project.isEnabled() && PROJECT.equals(project.qualifier()), "Invalid project key");
 
-      Collection<BranchDto> branches = dbClient.branchDao().selectByComponent(dbSession, project);
+      Collection<BranchDto> branches = dbClient.branchDao().selectByComponent(dbSession, project).stream()
+        .filter(b -> b.getBranchType() == SHORT || b.getBranchType() == LONG)
```
Comment (janos-ss) [inline server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java]: Maybe add a TODO to replace with condition that depends on key type, when it's ready.

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskFormatter.java [hunk 6]
```diff
@@ -144,12 +142,25 @@ public TaskFormatter(DbClient dbClient, System2 system2) {
...
+            builder.setPullRequestTitle(b);
```
Comment (janos-ss) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskFormatter.java]: What about the url and the branch name of a pull request?

Comment (teryk) [reply server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskFormatter.java]: not needed according to spec with @stas-vilchik 
do you confirm?

Comment (stas-vilchik) [reply server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskFormatter.java]: I confirm that neither branch name, nor url is needed.

**Pull request merged**